File: src/content/3.5/code/scala/snippet15.scala
implicit def writerMonad[W: Monoid] = new Monad[Writer[W, ?]] {
  def flatMap[A, B](wa: Writer[W, A])(k: A => Writer[W, B]): Writer[W, B] =
    wa match {
      case Writer((a, w)) =>
        val ((a1, w1)) = runWriter(k(a))
        Writer((a1, Monoid[W].combine(w, w1)))
    }

  def pure[A](a: A) = Writer(a, Monoid[W].empty)
}

File: src/content/3.5/code/ocaml/snippet15.ml
module WriterMonad (W : Monoid) : Monad_Bind
with type 'a t = (W.t, 'a) writer =
struct
  type 'a t = (W.t, 'a) writer

  let ( >>= ) (Writer (a, w)) k =
    let a', w' = run_writer (k a) in
    Writer (a', W.mappend w w')

  let return a = Writer (a, W.mempty)
end
